File: frontend/src/lib/types/journal.ts
// Mirrors backend/app/schemas/journal.py — keep in sync.

export type EntryPrecision = "exact" | "approximate" | "backfilled";
export type EntryStatus = "open" | "closed";

export type JournalEntry = {
  id: number;
  ticker: string;
  shares: number;
  cost_basis: number;
  price_per_share: number | null;

  entry_date: string | null;
  entry_precision: EntryPrecision;

  thesis: string | null;
  stop_loss: number | null;
  target_price: number | null;
  time_horizon: string | null;

  status: EntryStatus;
  exit_date: string | null;
  exit_price: number | null;
  exit_notes: string | null;

  notes: string | null;
  created_at: string;
  updated_at: string;

  // Derived
  current_price: number | null;
  market_value: number | null;
  unrealized_pnl: number | null;
  realized_pnl: number | null;
  pct_change: number | null;
};

export type JournalAggregates = {
  count: number;
  total_cost_basis: number;
  total_market_value: number | null;
  unrealized_pnl: number | null;
  pct_change: number | null;
};

export type OpenPositions = {
  positions: JournalEntry[];
  aggregates: JournalAggregates;
};

export type CreateEntryBody = {
  ticker: string;
  shares: number;
  cost_basis: number;
  entry_date?: string | null;
  entry_precision?: EntryPrecision;
  thesis?: string | null;
  stop_loss?: number | null;
  target_price?: number | null;
  time_horizon?: string | null;
  notes?: string | null;
};

export type UpdateEntryBody = Partial<
  Omit<CreateEntryBody, "ticker"> & {
    ticker?: never; // can't change ticker
  }
>;

export type CloseEntryBody = {
  exit_price: number;
  exit_date?: string | null;
  exit_notes?: string | null;
};

File: backend/app/schemas/journal.py
"""Journal schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


EntryPrecision = Literal["exact", "approximate", "backfilled"]
EntryStatus = Literal["open", "closed"]


class JournalEntryOut(BaseModel):
    id: int
    ticker: str
    shares: float
    cost_basis: float
    price_per_share: float | None = None

    entry_date: datetime | None = None
    entry_precision: EntryPrecision

    thesis: str | None = None
    stop_loss: float | None = None
    target_price: float | None = None
    time_horizon: str | None = None

    status: EntryStatus
    exit_date: datetime | None = None
    exit_price: float | None = None
    exit_notes: str | None = None

    notes: str | None = None
    created_at: datetime
    updated_at: datetime

    # Derived fields (populated for open/closed positions respectively)
    current_price: float | None = None
    market_value: float | None = None
    unrealized_pnl: float | None = None
    realized_pnl: float | None = None
    pct_change: float | None = None


class JournalAggregates(BaseModel):
    count: int
    total_cost_basis: float
    total_market_value: float | None = None
    unrealized_pnl: float | None = None
    pct_change: float | None = None


class OpenPositionsOut(BaseModel):
    positions: list[JournalEntryOut]
    aggregates: JournalAggregates


class CreateEntryRequest(BaseModel):
    ticker: str = Field(min_length=1, max_length=10)
    shares: float = Field(gt=0)
    cost_basis: float = Field(gt=0)
    entry_date: datetime | None = None
    entry_precision: EntryPrecision = "exact"
    thesis: str | None = None
    stop_loss: float | None = Field(default=None, gt=0)
    target_price: float | None = Field(default=None, gt=0)
    time_horizon: str | None = None
    notes: str | None = None

    @field_validator("ticker")
    @classmethod
    def _upper(cls, v: str) -> str:
        return v.strip().upper()


class UpdateEntryRequest(BaseModel):
    shares: float | None = Field(default=None, gt=0)
    cost_basis: float | None = Field(default=None, gt=0)
    entry_date: datetime | None = None
    entry_precision: EntryPrecision | None = None
    thesis: str | None = None
    stop_loss: float | None = None
    target_price: float | None = None
    time_horizon: str | None = None
    notes: str | None = None


class CloseEntryRequest(BaseModel):
    exit_price: float = Field(gt=0)
    exit_date: datetime | None = None
    exit_notes: str | None = None
